### PR TITLE
feat: Snap account async transaction redirect

### PR DESCRIPTION
**Pull request by owencraston**
## **Description**

Progresses: https://github.com/MetaMask/accounts-planning/issues/35

This PR implements the `redirectUser` callback that was added in the [eth-snap-keyring](https://github.com/MetaMask/eth-snap-keyring/pull/136). This callbacks allows us to redirect the user to a given url or direct them with custom instructions to sign an async transaction.

[Design](https://www.figma.com/file/K0csZegZ5XQ6TjfKk3n4Z8/Add-dialog-to-inform-about-redirection-to-snap-site-%2335?type=design&node-id=604-2586&mode=design&t=tRpxcs4qeRiRKxPl-0)

## **Manual testing steps**

1. checkout this [branch](https://github.com/MetaMask/snap-simple-keyring/pull/98) in the snap simple keyring repo (this branch is wip but it will work for testing)
2. run `yarn install` and `yarn start`
3. checkout this branch and run `yarn install && yarn start --build-type flask`
4. in your browser load the dist folder into the extensions via `load unpacked`
5. setup a wallet in MM
6. 3. navigate to http://localhost:8000/
7. click `create account`  in the dapp site and approve the creation flow
8. toggle the `Use Synchronous Approval` button at the top of the page so that it is grey. This will turn on async approvals 
9. navigate to https://metamask.github.io/test-dapp/
10. connect your MM to the snap
11. scroll down and click `Personal Sign`
12. a pop up should open in MM, accept that sign
13. a new pop up should open that matches the [above designs](https://www.figma.com/file/K0csZegZ5XQ6TjfKk3n4Z8/Add-dialog-to-inform-about-redirection-to-snap-site-%2335?type=design&node-id=604-2586&mode=design&t=tRpxcs4qeRiRKxPl-0)

## **Screenshots/Recordings**

Blocked URL
Screenshot 2023-10-12 at 11 16 35 AM

redirect with message and URL
Screenshot 2023-10-12 at 11 14 55 AM

without message
Screenshot 2023-10-12 at 11 18 23 AM

without URL
Screenshot 2023-10-12 at 11 17 33 AM

### **Before**

N/A

### **After**

https://github.com/MetaMask/metamask-extension/assets/22918444/05213130-ac5d-4d65-ac10-2fe32a96115a

## **Related issues**
https://github.com/MetaMask/accounts-planning/issues/35

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained:
  - [ ] What problem this PR is solving.
  - [ ] How this problem was solved.
  - [ ] How reviewers can test my changes.
- [ ] I’ve indicated what issue this PR is linked to: Fixes #???
- [ ] I’ve included tests if applicable.
- [ ] I’ve documented any added code.
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by owencraston** [comment]
@metamaskbot update-policies

**Review by HowardBraham — DISMISSED** [review]

File: app/scripts/lib/snap-keyring/snap-keyring.ts
Comment (FrederikBolding) [inline app/scripts/lib/snap-keyring/snap-keyring.ts:60]: Could we check the length instead?

Also in what case would these not be defined? Do you not need to specify a URL to redirect? 🤔

File: ui/pages/confirmation/templates/snap-account-redirect.js
Comment (FrederikBolding) [inline ui/pages/confirmation/templates/snap-account-redirect.js:6]:
```suggestion
    if (url !== undefined && url !== null && url.length > 0) {
```

File: ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx
Comment (FrederikBolding) [inline ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:84]: Same applies here 🤔

File: ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx
Comment (FrederikBolding) [inline ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: I assume this shouldn't render when `isBlocked`? That doesn't seem to be the case

File: app/scripts/lib/snap-keyring/snap-keyring.ts
Comment (owencraston) [reply app/scripts/lib/snap-keyring/snap-keyring.ts:60]: @FrederikBolding the url and message come from the [eth-snap-keyring](https://github.com/MetaMask/eth-snap-keyring/pull/136) which is sent from a snap ([example](https://github.com/MetaMask/snap-simple-keyring/pull/98/files#diff-22aa27652c8dffcd76652c4ed22b9e646ab8cc3b43594d0eaa303e272017a367R219-R221)) so they can never be undefined but they can be empty strings. we can handle the case where one of them is not set but when both are empty we wont do anything. See the [designs ](https://www.figma.com/file/K0csZegZ5XQ6TjfKk3n4Z8/Add-dialog-to-inform-about-redirection-to-snap-site-%2335?type=design&node-id=604-2586&mode=design&t=tRpxcs4qeRiRKxPl-0)

File: ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx
Comment (owencraston) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: based on these designs we still render the message and/or url when it is blocked but we also render a banner alert. One point of confusion here might be that the `SnapAccountRedirectMessage` component also renders the URL.
Screenshot 2023-10-11 at 11 41 28 AM

**Comment by owencraston** [comment]
Once we merge [this](https://github.com/MetaMask/snap-simple-keyring/pull/98) change into the Snap Simple Keyring we can create a new release and fix the broken e2e test

File: ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx
Comment (FrederikBolding) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: Are we sure about showing blocked URLs to the user? 🤔 

The snaps team is about to ship a feature to allow links in custom UIs and we were planning to not show URLs if they are in the phishing list.

File: app/scripts/lib/snap-keyring/snap-keyring.ts
Comment (FrederikBolding) [reply app/scripts/lib/snap-keyring/snap-keyring.ts:60]: are they asserted somewhere to guarantee that they are strings? If so, I think checking the length is slightly cleaner than checking against `''`

File: ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx
Comment (eriknson) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: Do we have a link to these designs?

Comment (owencraston) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: @eriknson https://www.figma.com/file/K0csZegZ5XQ6TjfKk3n4Z8/Add-dialog-to-inform-about-redirection-to-snap-site-%2335?type=design&node-id=604-2586&mode=design

Comment (owencraston) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: cc @raulmono

Comment (raulmono) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: Thanks for the heads up – not displaying the URL when it's blocked sounds good. It might look like this: 
image

Comment (owencraston) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: @raulmono would we still want to show the `Go to site` button?

Comment (raulmono) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: @owencraston yeah, you're right – we should show that either

Comment (raulmono) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: wdyt? @eriknson @FrederikBolding

Comment (FrederikBolding) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: Sounds good to me 👍

Comment (owencraston) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:85]: Updated to reflect that change
Screenshot 2023-10-11 at 3 22 58 PM

File: app/scripts/lib/snap-keyring/snap-keyring.ts
Comment (owencraston) [reply app/scripts/lib/snap-keyring/snap-keyring.ts]: we convert all the values to '' if they are null [here](https://github.com/MetaMask/eth-snap-keyring/pull/136/files#diff-6c3541c9e03e286cae318f6482c33d1a808662e60c02c91ea5dea83b28e91bb1R265). That being said I added the length check in the latest commit.

**Comment by plasmacorral** [comment]
URL is truncated and the user has no way to verify where they are agreeing to go.  Can we offer to present the entire URL if the user hovers on it?  Can we also support allowing the user to right click and copy the link?

Observed in f26bd122dfc9fc41e41a26a9ee6867019d7728bd on Chrome 117 in Mac Sonoma 14.0

<div style="display: inline-block"> </div>

**Comment by plasmacorral** [comment]
Observed in https://github.com/MetaMask/metamask-extension/commit/f26bd122dfc9fc41e41a26a9ee6867019d7728bd on Chrome 117 in Mac Sonoma 14.0

Running into a race condition here where sometimes the pop-up closes while showing the yellow spinning wheel and then the redirect is opened, but other times the redirect just flashes underneath the spinning wheel.  When I see the flash and get stuck with a yellow spinnning wheel presented, if I close the pop-up view from there the pending tx is listed as a failure.  However, I can still complete the flow if I tap the extension icon and open another pop-up which shows me an unobstructed redirect.

Reproduction steps:

1. Have ganache running locally with network ID 1
2. Setup flask with ganache SRP
3. create a snap account and fund it w tokens from account 1
4. open test dapp and connect
5. Deploy piggy back contract
6. Use the Piggy bank contract> deposit repeatedly

Video of deposit working as desired in pop-up view: https://recordit.co/oxYKxlKq7g
Video of opening a second pop-up and completing flow (workaround): https://recordit.co/mRZeJkVyek
Video of transaction showing as failed in activity after user closes pop-up: https://recordit.co/E18rZ1IzDq

**Comment by plasmacorral** [comment]
We may need to further explore the proposal discussed in slack wrt the loading indicator: https://consensys.slack.com/archives/C05FUV4205C/p1696612858307329?thread_ts=1696610011.723579&cid=C05FUV4205C

Otherwise the behavior in expanded view will be challenging.  Basically, the user gets stuck with the yellow spinner until they click the extension icon to open pop-up view to see the redirect and complete the flow.

https://recordit.co/4KdIloMvA3

**Comment by danroc** [comment]
IIRC, the Figma had a fixed-size box for the URL, but from the screenshots it seems that it's adapting to the URL's length.

**Comment by plasmacorral** [comment]
If I lock extension while I have a transaction pending approval in the snap, after I unlock the wallet I am not able to approve the pending tx in the snap.  Further, once I create this condition all subsequent transaction attempts are not producing a request in the snap even though it continues to show the synchronous approval toggle in the off position.

Reproduction steps:
1. Have Ganache running locally with chain ID set to 1
2. Configure flask with ganache SRP
3. Connect and install to SSK
4. Create a new snap account
5. fund snap account w tokens from account 1
6. open test dapp and connect
7. Scroll down and attempt to deploy piggy back contract (or can just do a p2p send if you prefer)
8. If you see redirect notice tap the cancel CTA, if you are stuck at the yellow spinning wheel click the extension icon and then cancel the redirect
9. Tap kebab menu
10. Select `Lock MetaMask`
11. At prompt Unlock MetaMask
12. Open SSK and tap `List requests`
13. Note a request does appear for the tx from step 7
14. Attempt to approve that request ID
15. Note that approval fails `request xxxxx not found`
16. tap `List requests` and note nothing pending
17. Attempt another tx and confirm in extension
18. Note that new request never makes it to dapp, still shows empty array on list requests

Recording: https://recordit.co/MkV2IJddQD

**Comment by owencraston** [comment]
> IIRC, the Figma had a fixed-size box for the URL, but from the screenshots it seems that it's adapting to the URL's length.

@danroc based on @plasmacorral feedback I am proposing to change this

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Comment by montelaidev** [comment]
@metamaskbot update-policies

**Review by Jordanwaslistening — APPROVED** [review]

**Comment by eriknson** [comment]
> @plasmacorral @raulmono I updated the designs to wrap the URL if it is longer than the textbox to address mikes valid security concerns Screenshot 2023-10-11 at 8 51 46 PM

- Header's shadow is broken / missing
- Border color on the external link component is different from delineator border color
- Icon for the external link component is wrong, should be `export`
- Font-size for the delineator body text should be `body-md (baseline)` (right now it looks like `sm`)
- Horizontal margin between popup container and delineator looks off, should be `16px` all around

**Comment by montelaidev** [comment]
> > @plasmacorral @raulmono I updated the designs to wrap the URL if it is longer than the textbox to address mikes valid security concerns Screenshot 2023-10-11 at 8 51 46 PM
> 
> * Header's shadow is broken / missing
> * Border color on the external link component is different from delineator border color
> * Icon for the external link component is wrong, should be `export`
> * Font-size for the delineator body text should be `body-md (baseline)` (right now it looks like `sm`)
> * Horizontal margin between popup container and delineator looks off, should be `16px` all around

Made the changes in this commit 9246540c13b8dd1c881c6f02871b32fe114f080f

**Comment by HowardBraham** [comment]
Why did the privacy-snapshot.json add `aus5.mozilla.org`?

**Comment by montelaidev** [comment]
> Why did the privacy-snapshot.json add `aus5.mozilla.org`?

it was complaining about this link during the firefox e2e tests

File: ui/pages/snap-account-redirect/components/url-display-box.tsx
Comment (georgewrmarshall) [inline ui/pages/snap-account-redirect/components/url-display-box.tsx:39]: Is this component meant to look like a TextField? I'm trying to understand the intention behind displaying a link in a TextField?

Screenshot 2023-10-12 at 1 58 38 PM

In the design it uses a TextField so the intention would be for the value to be editable or readonly which should allow the user to copy the value easily, I haven't seen a pattern of a TextField being used for a link?

Screenshot 2023-10-12 at 1 59 07 PM

File: ui/pages/snap-account-redirect/snap-account-redirect.stories.tsx
Comment (georgewrmarshall) [inline ui/pages/snap-account-redirect/snap-account-redirect.stories.tsx:18]: You can make the story more interactive by moving the data to args and allowing the controls to be used

```suggestion
  argTypes: {
    snapName: {
      control: {
        type: 'text',
      },
    },
    isBlockedUrl: {
      control: {
        type: 'boolean',
      },
    },
    url: {
      control: {
        type: 'text',
      },
    },
    message: {
      control: {
        type: 'text',
      },
    },
  },
  args: {
    snapName: 'Snap Simple Keyring',
    isBlockedUrl: false,
    url: 'https://metamask.github.io/snap-simple-keyring/0.2.4/',
    message: 'Redirecting to Snap Simple Keyring',
  },
} as Meta<typeof SnapAccountRedirect>;

export const DefaultStory = (args) => <SnapAccountRedirect {...args} />;

DefaultStory.storyName = 'Default';

```

**Review by georgewrmarshall — COMMENTED** [review]
Looking good! It would greatly speed up review of UI if there were stories for each file that included UI. Left some design questions for @raulmono

File: ui/pages/snap-account-redirect/components/url-display-box.tsx
Comment (eriknson) [reply ui/pages/snap-account-redirect/components/url-display-box.tsx:39]: Agreed, I think we should have another design pattern for (external) links. Or at the very least make it a `disabled` text field so it doesn't look editable

File: ui/pages/snap-account-redirect/snap-account-redirect.stories.tsx
Comment (montelaidev) [reply ui/pages/snap-account-redirect/snap-account-redirect.stories.tsx:18]: Changed on 523c7575f9119e48914b5059d4e2b53f6c5b2a3f

File: ui/pages/snap-account-redirect/components/redirect-url-icon.tsx
Comment (montelaidev) [reply ui/pages/snap-account-redirect/components/redirect-url-icon.tsx]: Changed on 976dfdfd19539715a2670b04881827a6c02985a0

File: ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx
Comment (montelaidev) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-context.tsx:97]: These are sub components of the SnapAccountRedirect and there is already a story for it.

File: ui/pages/snap-account-redirect/components/snap-account-redirect-message.tsx
Comment (montelaidev) [reply ui/pages/snap-account-redirect/components/snap-account-redirect-message.tsx:36]: These are sub components of the SnapAccountRedirect and there is already a story for it.

File: ui/pages/snap-account-redirect/components/url-display-box.tsx
Comment (montelaidev) [reply ui/pages/snap-account-redirect/components/url-display-box.tsx:39]: The url component right now is just a `Text` component and not a `TextField`. I agree, we should have another pattern for external links.

**Comment by danroc** [comment]
Hey @plasmacorral, [this](https://github.com/MetaMask/metamask-extension/pull/21312#issuecomment-1758719516) is a bug! It's not related to the confirmation screen, we can also reproduce it with hardware wallets.

We will need to address this for all wallets.

**Comment by owencraston** [comment]
I just updated the PR description with the latest UX changes.

File: ui/pages/snap-account-redirect/components/url-display-box.tsx
Comment (owencraston) [reply ui/pages/snap-account-redirect/components/url-display-box.tsx:39]: I migrated from a TextField to a normal text component but I agree we should find a better more standard pattern. This is something that should happen at a later date IMO.

File: app/scripts/lib/snap-keyring/snap-keyring.ts
Comment (FrederikBolding) [inline app/scripts/lib/snap-keyring/snap-keyring.ts:62]: What is the reasoning for showing the popup if the URL is blocked anyway? 🤔

File: app/scripts/lib/snap-keyring/snap-keyring.ts
Comment (owencraston) [reply app/scripts/lib/snap-keyring/snap-keyring.ts:61]: There are cases where there are no redirect URLs but there are instructions offered on how to complete the signing. Examples of this are hardware wallet snaps are the silent laboratories because they use a companion app.

File: app/scripts/lib/snap-keyring/snap-keyring.ts
Comment (owencraston) [reply app/scripts/lib/snap-keyring/snap-keyring.ts:62]: It is another opportunity for us to inform the user of potential malicious activity.

File: app/scripts/lib/snap-keyring/snap-keyring.ts
Comment (FrederikBolding) [reply app/scripts/lib/snap-keyring/snap-keyring.ts:61]: Okay, one nit then, perhaps in that case the copy shouldn't be "cancel"? Seems confusing if the message is actually useful? Perhaps it should just be OK?

File: test/e2e/accounts/test-snap-accounts.spec.js
Comment (owencraston) [reply test/e2e/accounts/test-snap-accounts.spec.js:514]: There is a bug in the send flow that loads an activity indicator until the request is approved. In the case of the async transactions they will remain loading until the modal is closed. This is a larger scale bug that is effecting hardware wallets as well and there are already other issues to fix this. But to get around it we need to open and close the modal.

File: app/scripts/lib/snap-keyring/snap-keyring.ts
Comment (FrederikBolding) [reply app/scripts/lib/snap-keyring/snap-keyring.ts:62]: Okay, one nit then, should we disable showing the "follow the instructions..." copy then. That doesn't seem useful if the instructions are malicious?

Comment (owencraston) [reply app/scripts/lib/snap-keyring/snap-keyring.ts:62]: I made the update in the latest change
Screenshot 2023-10-12 at 10 48 01 AM

Comment (owencraston) [reply app/scripts/lib/snap-keyring/snap-keyring.ts:62]: @FrederikBolding I also changed the cancel button to say close. This matches the copy in [these](https://www.figma.com/file/K0csZegZ5XQ6TjfKk3n4Z8/Add-dialog-to-inform-about-redirection-to-snap-site-%2335?type=design&node-id=604-2586&mode=design&t=XHzW2Gm7AXUtIUhL-0) designs and makes more sense

**Review by FrederikBolding — DISMISSED** [review]

**Review by FrederikBolding — APPROVED** [review]

**Review by gantunesr — APPROVED** [review]

**Review by danjm — APPROVED** [review]

**Pull request merged**
